In [ ]:
#Social_Networks_Ads.csv

**Importing the libraries**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report, f1_score
from sklearn.model_selection import cross_val_score

**Importing the dataset**

In [ ]:
df = pd.read_csv("/content/Social_Network_Ads.csv")

In [ ]:
df

User ID  Gender  Age  EstimatedSalary  Purchased
0    15624510    Male   19            19000          0
1    15810944    Male   35            20000          0
2    15668575  Female   26            43000          0
3    15603246  Female   27            57000          0
4    15804002    Male   19            76000          0
..        ...     ...  ...              ...        ...
395  15691863  Female   46            41000          1
396  15706071    Male   51            23000          1
397  15654296  Female   50            20000          1
398  15755018    Male   36            33000          0
399  15594041  Female   49            36000          1

[400 rows x 5 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   User ID          400 non-null    int64 
 1   Gender           400 non-null    object
 2   Age              400 non-null    int64 
 3   EstimatedSalary  400 non-null    int64 
 4   Purchased        400 non-null    int64 
dtypes: int64(4), object(1)
memory usage: 15.8+ KB


In [ ]:
df.isna().sum()

User ID            0
Gender             0
Age                0
EstimatedSalary    0
Purchased          0
dtype: int64

In [ ]:
# data cleaning
df = df.drop("User ID",axis=1)

In [ ]:
# encoding
le = LabelEncoder()
gender_le = le.fit_transform(df["Gender"])
df.Gender = gender_le

In [ ]:
df.head()

Gender  Age  EstimatedSalary  Purchased
0       1   19            19000          0
1       1   35            20000          0
2       0   26            43000          0
3       0   27            57000          0
4       1   19            76000          0

**Splitting the dataset into the Training set and Test set**

In [ ]:
X = df.iloc[:,:3].values
y = df.iloc[:,-1].values

In [ ]:
X.shape,y.shape

((400, 3), (400,))

In [ ]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2)

**Feature Scaling**

In [ ]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

**Fitting K-NN to the Training set**

In [ ]:
# function for finding the best k value
def best_k(data, k):
  x = df[['Gender','Age','EstimatedSalary']].values 
  y = df['Purchased'].values
  clf = KNeighborsClassifier(n_neighbors=k) 
  clf.fit(x, y) 
  print("K : ", k, " Train Accuracy : ", clf.score(x,y), " Val Accuracy : ", np.mean(cross_val_score(clf, x, y, cv=5))) 

In [ ]:
for k in [1,2,3,4,5,6,7,8,9,10,20]:
  best_k(df,k)

K :  1  Train Accuracy :  0.9975  Val Accuracy :  0.7849999999999999
K :  2  Train Accuracy :  0.9  Val Accuracy :  0.7575000000000001
K :  3  Train Accuracy :  0.8975  Val Accuracy :  0.76
K :  4  Train Accuracy :  0.8625  Val Accuracy :  0.7699999999999999
K :  5  Train Accuracy :  0.8725  Val Accuracy :  0.7674999999999998
K :  6  Train Accuracy :  0.8625  Val Accuracy :  0.7625
K :  7  Train Accuracy :  0.87  Val Accuracy :  0.7675000000000001
K :  8  Train Accuracy :  0.85  Val Accuracy :  0.775
K :  9  Train Accuracy :  0.8675  Val Accuracy :  0.7649999999999999
K :  10  Train Accuracy :  0.845  Val Accuracy :  0.75
K :  20  Train Accuracy :  0.8175  Val Accuracy :  0.7625


* best k value occurs at k=1

In [ ]:
# applying the best k value
model = KNeighborsClassifier(n_neighbors=1)
model.fit(X_train,y_train)

KNeighborsClassifier(n_neighbors=1)

**Predicting the Test set results**

In [ ]:
y_pred = model.predict(X_test)

**Making the Confusion Matrix**

In [ ]:
print(confusion_matrix(y_test,y_pred))

[[51  4]
 [ 2 23]]


interpretation of confusion matrix:
* true positive = 51
* false positive = 4
* false negative = 2
* true negative = 23

In [ ]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.96      0.93      0.94        55
           1       0.85      0.92      0.88        25

    accuracy                           0.93        80
   macro avg       0.91      0.92      0.91        80
weighted avg       0.93      0.93      0.93        80



interpretation:
  * f1 score of 93% shows that the model works very well.